src/dragonball/src/dbs_pci/src/bus.rs

```diff
@@ -0,0 +1,415 @@
+        for req in constraints {
+            match req {
+                ResourceConstraint::PioAddress { range, align, size } => {
+                    let mut constraint = Constraint::new(*size as u64).align(*align as u64);
+                    if let Some((min, max)) = range {
+                        constraint = constraint.min(*min as u64).max(*max as u64);
+                    }
+                    match state.ioport_resources.allocate(&constraint) {
+                        Some(range) => {
+                            resources.append(Resource::PioAddressRange {
+                                base: range.min as u16,
+                                size: range.len() as u16,
+                            });
+                        }
+                        None => {
+                            Self::free_all_resource(state, resources);
+                            return Err(Error::NoResources);
+                        }
+                    }
+                }
+                ResourceConstraint::MmioAddress { range, align, size } => {
+                    let mut constraint = Constraint::new(*size).align(*align);
+                    if let Some((min, max)) = range {
+                        constraint = constraint.min(*min).max(*max);
+                    }
+                    match state.iomem_resources.allocate(&constraint) {
+                        Some(range) => {
+                            resources.append(Resource::MmioAddressRange {
+                                base: range.min,
+                                size: range.len(),
+                            });
+                        }
+                        None => {
+                            Self::free_all_resource(state, resources);
+                            return Err(Error::NoResources);
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        Ok(resources)
+    }
+
+    /// Free allocated PCI IO resources.
+    pub fn free_resources(&self, resources: DeviceResources) {
+        // Don't expect poisoned lock here.
+        let state = self.state.write().unwrap();
+        Self::free_all_resource(state, resources);
+    }
+
+    // free all device resource under pci bus
+    fn free_all_resource(mut state: RwLockWriteGuard<PciBusContent>, resources: DeviceResources) {
+        for res in resources.get_all_resources().iter() {
+            match res {
+                Resource::PioAddressRange { base, size } => {
+                    let range = Range::new(*base as u64, (*base + *size - 1) as u64);
+                    state.ioport_resources.free(&range);
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    let range = Range::new(*base, *base + *size - 1);
+                    state.iomem_resources.free(&range);
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+impl PartialEq for PciBus {
```

Comment:
Do we really need these code at this moment?

Reply:
nope, but this will be used in later PRs and it's basic functionality for PciBus.